src/charm.py

```diff
@@ -400,122 +398,21 @@ def _on_vault_kv_client_detached(self, event: VaultKvClientDetachedEvent):
-        try:
-            vault.make_latest_pki_issuer_default(mount=PKI_MOUNT)
-        except VaultClientError as e:
-            logger.error("Failed to make latest issuer default: %s", e)
-
-    def _intermediate_ca_exceeds_role_ttl(
-        self, vault: VaultClient, intermediate_ca_certificate: Certificate
-    ) -> bool:
-        """Check if the intermediate CA's remaining validity exceeds the role's max TTL.
-
-        Vault PKI enforces that issued certificates cannot outlast their signing CA.
-        This method ensures that the intermediate CA's remaining validity period
-        is longer than the maximum TTL allowed for certificates issued by this role.
-        """
-        current_ttl = vault.get_role_max_ttl(role=PKI_ROLE_NAME, mount=PKI_MOUNT)
-        if (
-            not current_ttl
-            or not intermediate_ca_certificate.expiry_time
-            or not intermediate_ca_certificate.validity_start_time
-        ):
-            return False
-        certificate_validity = (
-            intermediate_ca_certificate.expiry_time
-            - intermediate_ca_certificate.validity_start_time
-        )
-        certificate_validity_seconds = certificate_validity.total_seconds()
-        return certificate_validity_seconds > current_ttl
-
-    def _calculate_pki_certificates_ttl(self, certificate: Certificate) -> int:
-        """Calculate the maximum allowed validity of certificates issued by PKI.
-
-        Return half the CA certificate validity in seconds.
-        """
-        if not certificate.expiry_time or not certificate.validity_start_time:
-            raise ValueError("Invalid CA certificate with no expiry time or validity start time")
-        ca_validity_time = certificate.expiry_time - certificate.validity_start_time
-        ca_validity_seconds = ca_validity_time.total_seconds()
-        return int(ca_validity_seconds / 2)
-
-    def _get_pki_intermediate_ca(
-        self,
-    ) -> Tuple[ProviderCertificate | None, PrivateKey | None]:
-        """Get the intermediate CA certificate."""
-        certificate_request = self._get_certificate_request()
-        if not certificate_request:
-            logger.error("Certificate request is not valid")
-            return None, None
-        provider_certificate, private_key = self.tls_certificates_pki.get_assigned_certificate(
-            certificate_request=certificate_request
+        manager = PKIManager(
+            self,
+            vault,
+            common_name,
+            PKI_MOUNT,
+            PKI_ROLE_NAME,
+            self.vault_pki,
+            tls_certificates_pki=self.tls_certificates_pki,
         )
-        if not provider_certificate:
-            logger.debug("No intermediate CA certificate available")
-            return None, None
-        if not private_key:
-            logger.debug("No private key available")
-            return None, None
-        return provider_certificate, private_key
+        manager.configure()
```

Comment:
I'm a bit confused by the fact that there is both a `configure` and a `sync` methods in this manager. Is this really necessary?

Reply:
see https://github.com/canonical/vault-k8s-operator/pull/559#discussion_r1890166870